ethcore/src/state/mod.rs

```diff
@@ -539,8 +542,49 @@ impl<B: Backend> State<B> {
 			|a| a.as_ref().and_then(|account| account.storage_root().cloned()))
 	}
 
-	/// Mutate storage of account `address` so that it is `value` for `key`.
-	pub fn storage_at(&self, address: &Address, key: &H256) -> TrieResult<H256> {
+	/// Get the value of storage at last checkpoint.
+	pub fn checkpoint_storage_at(&self, checkpoint_index: usize, address: &Address, key: &H256) -> TrieResult<Option<H256>> {
+		if let Some(ref checkpoint) = self.checkpoints.borrow().get(checkpoint_index) {
+			Ok(Some(if let Some(entry) = checkpoint.get(address) {
+				match entry {
+					Some(entry) => {
+						match entry.account {
+							Some(ref account) => {
+								if let Some(value) = account.cached_storage_at(key) {
+									value
+								} else {
+									// This key has checkpoint entry, but not in the entry's cache.
+									// So it's not modified at all.
+									self.original_storage_at(address, key)?
+								}
+							},
+							None => {
+								// The account didn't exist at that point. Return empty value.
+								H256::new()
+							},
+						}
+					},
+					None => {
+						// The value was not cached at that checkpoint, meaning it was not modified.
+						self.original_storage_at(address, key)?
+					},
+				}
+			} else {
+				// This key does not have a checkpoint entry. We use the current value.
```

Comment:
Is it possible here that the checkpoint do not have the value because of the indexed checkpoint mechanism (I am unsure about the indexed checkpoint mechanism seeing that it was added in a later commit, I am wondering), leading to issues?

Reply:
Yeah the logic seems to be flawed there. My current fix is recursive, and I can't think of a better way right now using the current checkpointing method.

I think we need to finish the state mod refactoring before this PR lands.

Reply:
I realized that I [was](https://github.com/paritytech/parity-ethereum/pull/9319#discussion_r213246747) doing the index change in wrong direction. Just fixed this. Current solution is still recursive, but it has an upper bound. So I think this would work fine without #9427. However, #9427 would still be a really nice optimization.